cluster_management/src/main/java/com/pinterest/rocksplicator/CurrentStatesConfigGenerator.java

```diff
@@ -0,0 +1,505 @@
+ * the instance Participant is a leader. However, to use this class and guarantee that only
+ * one instance is active throughout Participant cluster, we need a notification on when the
+ * leadership is acquired and when is it released. This is because we manage the notification
+ * from helix directly through use of RoutingTableProvider. Hence even if helixManager is shared
+ * between the object of this class and the HelixCustomCodeRunner, the listeners of
+ * RoutingTableProvider will not be unregistered when the leadership is lost.
+ *
+ * Hence is is recommended that this class be only used in either Spectator as a Stand Alone mode
+ * or the DistributedSpectator which provides hooks for leadership lost v/s leadership acquired.
+ */
+public class CurrentStatesConfigGenerator
+    implements RoutingTableChangeListener,
+               ConfigGeneratorIface,
+               ExternalViewChangeListener,
+               LiveInstanceChangeListener,
+               InstanceConfigChangeListener,
+               ResourceConfigChangeListener,
+               IdealStateChangeListener {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ConfigGenerator.class);
+  private final String clusterName;
+  private final Map<String, String> hostToHostWithDomain;
+  private final HelixManager helixManager;
+  private final HelixAdmin helixAdmin;
+  private final ReentrantLock synchronizedCallbackLock;
+  private final ShardMapPublisher shardMapPublisher;
+  private final RocksplicatorMonitor monitor;
+  private final ConcurrentMap<String, IdealState> idealStateMap;
+  private ExternalViewLeaderEventLogger externalViewLeaderEventLogger;
+  private RoutingTableProvider routingTableProvider;
+
+  public CurrentStatesConfigGenerator(
+      final String clusterName,
+      final HelixManager helixManager,
+      final ShardMapPublisher<JSONObject> shardMapPublisher,
+      final RocksplicatorMonitor monitor,
+      final ExternalViewLeaderEventLogger externalViewLeaderEventLogger) throws Exception {
+    this.clusterName = clusterName;
+    this.helixManager = helixManager;
+    this.helixAdmin = helixManager.getClusterManagmentTool();
+    this.shardMapPublisher = shardMapPublisher;
+    this.hostToHostWithDomain = new HashMap<String, String>();
+    this.monitor = monitor;
+    this.synchronizedCallbackLock = new ReentrantLock();
+    this.externalViewLeaderEventLogger = externalViewLeaderEventLogger;
+
+    this.idealStateMap = new ConcurrentHashMap<>();
+
+    this.routingTableProvider =
+        new RoutingTableProvider(this.helixManager, PropertyType.CURRENTSTATES);
+
+    this.routingTableProvider.addRoutingTableChangeListener(this, this);
+
+    helixManager.addResourceConfigChangeListener(this);
+    helixManager.addIdealStateChangeListener(this);
+    helixManager.addExternalViewChangeListener(this);
+    helixManager.addLiveInstanceChangeListener(this);
```

Comment:
Remove this... not needed as RoutingTable provides this from local cache populated in background.